[DEEP WIP] stroke and fill

stalgiag opened the pull request:
Hi,

So I tried to roll back a bunch of changes and simplify everything as much as possible with this one so that I can face this very specific problem before moving along. I am actually preventing the library from doing a lot of stuff it would do otherwise just for the sake of simplicity. I realized that a big thing that was happening that was causing me to only be able to draw a portion of the lineVertices was that I was being held back by the number of items in the normal and uv buffers. I have a version where I more or less fix that but it is a bit distracting from the current problem which has more to do with the order that the lineVertices are connecting to each other in.

I tried to clearly indicate what I was doing as much as possible using the obnoxious-but-visible /**COMMENTING STYLE**/. Also note that this won't build with the style checks and tests for obvious reasons but it should be fine for our purposes using `grunt watch:quick`. You can see what is happening with this by running the manual test in webgl/material/edgestroke. If for any reason you want to test a primitive other than a box you can do so by taking out the call to `.computeFaces()` towards the end of that specific primitives function call in the primitives.js file and adding `._makeTriangleEdges()` and `._edgesToVertices()` in its place.

Let me know if there is anything I can do to this branch to help clarify what is happening. I will continue working on it from my end but I am pretty firmly stuck at the moment.

Commit:
Merge branch 'processing/WebGL_GSOC17' into stroke_and_fill

Commit:
committing my testing version

Commit:
adding small example of another approach to build edges

Commit:
small change

kjhollen commented on src/webgl/p5.RendererGL.Retained.js:
```diff
@@ -50,32 +69,41 @@ p5.RendererGL.prototype.createBuffers = function(gId, obj) {
     3, gl.FLOAT, false, 0, 0);
-  gl.bindBuffer(gl.ARRAY_BUFFER, this.gHash[gId].vertexBuffer);
+  gl.bindBuffer(gl.ARRAY_BUFFER, this.gHash[gId].lineVertexBuffer);
   gl.bufferData(gl.ARRAY_BUFFER, data, gl.STATIC_DRAW);
```
Cool, I think it makes sense that these line vertices would be pushed to the aPosition attribute. It would make writing custom shaders easier by requiring the user to know less about our implementation, which is a good thing.

kjhollen replied:
One more related comment: I've been thinking about if the prep for the buffers could be abstracted enough to go into p5.Renderer.GL instead of in the separate immediate mode and retain mode files. I'm not yet convinced that the answer is yes, or that it's a good thing to do. I'm curious if you have any thoughts on this!

stalgiag replied:
I definitely think it would be possible to abstract it to that point. If a generic function was handling the buffer prepping it may have to take an unseemly number of arguments in order to be specific enough for all of the cases but I don't think that is really much of an issue and I think it would further the approach that the shader and texture class have been taking with which is a good approach imo. All of the work in that direction seems to really help with legibility.

This sounds fun and feasible. Would you mind me swiping the task?

kjhollen replied:
Not at all—go for it!!

kjhollen reviewed (approved):
This seems reasonable to me so far! I started a branch to see what might be needed for custom shaders now that we have the texture & shader objects. I think it makes sense for us to chat about the library's standard uniforms and attributes that would go into a shader so that people can write their own later (and know that the geometry & lights will link up properly). Let's chat tomorrow morning?

kjhollen committed:
fix ambient light, gl error when lighting is set but no texture

ambientLight was not incrementing the # of lights in the scene as it
should have been. must have been accidentally deleted during code cleanup
from a prior patch, oops!

and: importantly, restores the empty texture work Stalgia had added but
using the new p5.Texture architecture. a default texture is loaded
and used (pure white, 1x1 pixels from an in-memory p5.Image) when the
user has not supplied a texture to the shader.

kjhollen committed:
Merge pull request #2120 from kjhollen/lighting-texture-fixes

fix ambient light, gl error when lighting is set but no texture

sansumbrella commented:
Some debug drawing that might be helpful when looking at the cube. Any of these could be overlaid on the spinning cube:
- Draw the input vertices as points (the 8 vertices of the original cube geometry).
- Color the vertices based on their indexing
- Draw the output vertices as points

sansumbrella commented:
It would be helpful to see something other* than the cube on-screen at the same time. Just to give us more of an idea of what the lines are doing. In the case of the cube, it does mostly look like the issue is with some position indexing confusion, though it's not obvious how to grab the correct positions.

*like a 3d "ribbon" or stroked triangle in space

sansumbrella commented on src/webgl/p5.Geometry.js:
```diff
@@ -146,6 +150,98 @@ p5.Geometry.prototype.averagePoleNormals = function() {
+        if(this.vertices.length > i2)
+        {
+          this._addEdge(i1, i2, false, false);
+          this._addEdge(i2, i0, false, false);
```
Should this ever be `false, true`? Just curious that the second boolean parameter is always false across these calls. It could just be omitted if it is intended as a constant value.

stalgiag replied:
Totally! I realized that while working with Lauren yesterday. It isn't really part of the problem at hand because I am not even using those values yet but definitely worth noting.

sansumbrella commented on src/webgl/primitives.js:
```diff
@@ -151,13 +151,18 @@ p5.prototype.box = function(){
-      .computeNormals();
+      //.computeNormals(); NOT NECESSARY CURRENTLY
+      ._makeTriangleEdges()
+      ._edgesToVertices();
```
Is the intent to have two separate box geometries in the future, one for fill and one for stroke? Or will all the information be stored in one object?

stalgiag replied:
I think it would make sense to just store it in one object and send different values to the shader depending on which parts you want to draw but maybe they would need to be different geometries. Not sure yet but definitely worth considering.

Commit:
p5.vectors work with translate now

stalgiag committed:
Merge pull request #2124 from mlarghydracept/fix-for-p5.vector-in-translate

p5.vectors work with translate now

Commit:
Merge branch 'processing/WebGL_GSOC17' into stroke_and_fill
